panoply/datasource.py

```diff
@@ -7,24 +10,72 @@ class DataSource(events.Emitter):
     options = None
 
     # data source constructor
-    def __init__(self, source, options = {}):
-        super(DataSource, self).__init__()
-        
-        self.source = source
+    def __init__(self, source, options = {}, events = {}):
+        super(DataSource, self).__init__(events)
+
+        self.source = dict(source)
         self.options = options
 
+        if 'destination' not in source:
+            self.source['destination'] = source.get('type')
+
     # log a message
     def log(self, *msgs):
         if "logger" in self.options:
             self.options["logger"](msgs)
         else:
-            print msgs
+            print(msgs)
 
-    def progress(self, loaded, total, msg):
+    def progress(self, loaded, total, msg=''):
         self.fire("progress", {
             "loaded": loaded,
             "total": total,
             "msg": msg
         })
 
+# a decorator used to invalidate the access_token for oauth based data sources
+# this should be used on every method in the data source that fetches data
+# from the server (controlled by this oauth), and that needs to be invalidated
+def invalidate_token(refresh_url, callback=None,
+                     keys=['access_token', 'refresh_token']):
+    def _invalidate_token(f):
+        def wrapper(*args):
+            self = args[0]
+            try:
+                return f(*args)
+            except Exception, e:
+                try:
+                    self.log("Reinvalidating access_token...")
```

Comment:
`Refreshing` means that something is stale. Since I don't know if the access token is stale or not, I use `invalidating`